src/parser/shell_expand/words.rs

```diff
@@ -755,27 +806,71 @@ impl<'a> Iterator for WordIterator<'a> {
                     let output = &self.data[start..end];
                     self.read += 1;
-                    return Some(WordToken::Normal(output));
+                    return Some(WordToken::Normal(output,glob));
                 },
                 b'"' if self.flags & SQUOTE == 0 => {
                     self.flags ^= DQUOTE;
                     let end = if self.flags & EXPAND_PROCESSES == 0 { self.read+1 } else { self.read };
                     let output = &self.data[start..end];
                     self.read += 1;
-                    return Some(WordToken::Normal(output));
+                    return Some(WordToken::Normal(output,glob));
                 },
                 b' ' | b'{' if self.flags & (SQUOTE + DQUOTE) == 0 => {
-                    return Some(WordToken::Normal(&self.data[start..self.read]));
+                    return Some(WordToken::Normal(&self.data[start..self.read],glob));
+                },
+                b'$' | b'@' /*| b'['*/ if self.flags & SQUOTE == 0 => {
+                    return Some(WordToken::Normal(&self.data[start..self.read],glob));
+                },
+                b'[' if self.flags & SQUOTE == 0 => {
+                    let mut moves = 0;
+                    let mut square_bracket= 0;
+                    let mut rewind = true;
+                    let mut iter = iterator.clone();
+                    while let Some(character) = iter.next() {
+                        moves+=1;
+                        match character
+                        {
+                            b'[' => {
+                                square_bracket+=1;
+                            }
+                            b' '| b'"' | b'\'' | b'$' | b'{' | b'}' => {
+                                break;
+                            },
+                            b']' => {
+                                //we ignore the glob if it's smaller than 3, because [a] is a valid wild card and array
+                                //but the array meaning interpreted as a glob would actually be correct, whilst vice versa it wouldnt   
+                                if moves<=3 && square_bracket == 1 {
+                                }
+                                else{
+                                    rewind = false;
+                                    break;
+                                }
+                            }
+                            _=> (),
+                        }
+                    }
+                    if rewind == false{
+                        for _ in 0..moves {
+                            iterator.next();
+                        }
+                        self.read+=moves+1;
+                        glob = true;
+                    }
+                    else {
+                        return Some(WordToken::Normal(&self.data[start..self.read],glob));
+                    }
```

Comment:
Should be able to make the above a new function that returns a `Some(&'a str)` so it doesn't need to be copied. Then you can use the map method to convert the returned `Some(&'a str)` into the correct token that you need.